compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,362 @@
+//   }
+//   gpu.barrier
+struct LowerCoalescedGatherDMAPattern
+    : public OpRewritePattern<IREE::GPU::CoalescedGatherDMAOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LowerCoalescedGatherDMAPattern(MLIRContext *context,
+                                 ArrayRef<int64_t> workgroupSize,
+                                 int64_t subgroupSize,
+                                 int64_t maxLoadInstructionBits)
+      : OpRewritePattern<IREE::GPU::CoalescedGatherDMAOp>(context),
+        workgroupSize(workgroupSize), subgroupSize(subgroupSize),
+        maxLoadInstructionBits(maxLoadInstructionBits) {}
+
+  LogicalResult matchAndRewrite(IREE::GPU::CoalescedGatherDMAOp dmaOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = dmaOp.getLoc();
+
+    LDBG() << "=== matchAndRewrite called for: " << dmaOp;
+
+    // Check if this DMA op is eligible for transformation.
+    if (failed(isEligibleForGlobalDMA(dmaOp))) {
+      LDBG() << "Op is not eligible for transformation";
+      return failure();
+    }
+
+    scf::ForallOp forallOp = dmaOp->getParentOfType<scf::ForallOp>();
+
+    // Replace the entire scf.forall with amdgpu.gather_to_lds.
+    rewriter.setInsertionPoint(forallOp);
+
+    auto destMemRefType = dyn_cast<MemRefType>(dmaOp.getInit().getType());
+    int64_t numCopiesPerThread = getNumOfGlobalLoadsPerThread(
+        destMemRefType, subgroupSize, maxLoadInstructionBits);
+
+    Value indices = dmaOp.getIndices();
+    auto indicesType = cast<MemRefType>(indices.getType());
+
+    Value dest = dmaOp.getInit();
+    auto destType = cast<MemRefType>(dest.getType());
+
+    Value subgroupId = rewriter.create<gpu::SubgroupIdOp>(
+        loc, rewriter.getIndexType(), nullptr);
+    Value laneId =
+        rewriter.create<gpu::LaneIdOp>(loc, rewriter.getIndexType(), nullptr);
+
+    // Build a for loop skeleton.
+    Value lowerBound = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value upperBound =
+        rewriter.create<arith::ConstantIndexOp>(loc, numCopiesPerThread);
+    Value step = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    scf::ForOp forOp =
+        rewriter.create<scf::ForOp>(loc, lowerBound, upperBound, step);
+
+    // Now build the `scf.for` loop body.
+    auto delinearizeIndex = [&](Value index, ArrayRef<int64_t> shape) {
+      return rewriter
+          .create<affine::AffineDelinearizeIndexOp>(loc, index, shape)
+          .getMultiIndex();
+    };
+
+    auto getLinearizedGatherIndex = [&](Value sgIdVal, Value lIdVal,
+                                        Value indVar) -> Value {
+      int64_t numSubgroups = workgroupSize[0] / subgroupSize;
+      Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+      return rewriter.create<affine::AffineLinearizeIndexOp>(
+          loc, ValueRange{sgIdVal, indVar, lIdVal, zero},
+          ArrayRef<int64_t>{numSubgroups, numCopiesPerThread, subgroupSize,
```

Comment:
... Why are you scaling by the bitwidth? The index is _already_ in terms of the element type?